
diff --git a/internal/homekit/api.go b/internal/homekit/api.go
index d0a3c1c..2272367 100644
--- a/internal/homekit/api.go
+++ b/internal/homekit/api.go
@@ -173,5 +173,19 @@ func apiPairingHandler(w http.ResponseWriter, r *http.Request) {
 		}
 		api.ResponseJSON(w, pairingInfo)
 
+	case "DELETE":
+		query := r.URL.Query()
+		name := query.Get("name")
+		stream := query.Get("stream")
+		device_id := query.Get("device_id")
+		for _, s := range servers {
+			if name == s.mdns.Name || stream == s.stream || device_id == s.hap.DeviceID {
+				s.pairings = nil
+				s.UpdateStatus()
+				s.PatchConfig()
+				break;
+			}
+		}
+
 	}
 }
diff --git a/internal/homekit/homekit.go b/internal/homekit/homekit.go
index 86a5d4b..24b8a60 100644
--- a/internal/homekit/homekit.go
+++ b/internal/homekit/homekit.go
@@ -132,6 +132,8 @@ func Init() {
 			log.Error().Err(err).Caller().Send()
 		}
 	}()
+
+  discovery() // mdns scan
 }
 
 var log zerolog.Logger
diff --git a/internal/homekit/server.go b/internal/homekit/server.go
index cb114fe..7aa79f2 100644
--- a/internal/homekit/server.go
+++ b/internal/homekit/server.go
@@ -214,7 +214,11 @@ func (s *server) DelPair(conn net.Conn, id string) {
 			continue
 		}
 
-		s.pairings = append(s.pairings[:i], s.pairings[i+1:]...)
+		if strings.Contains(pairing, "permissions=1") {
+			s.pairings = nil
+		} else {
+			s.pairings = append(s.pairings[:i], s.pairings[i+1:]...)
+		}
 		s.UpdateStatus()
 		s.PatchConfig()
 		break
