
diff --git a/internal/homekit/homekit.go b/internal/homekit/homekit.go
index 24b8a60..04cc513 100644
--- a/internal/homekit/homekit.go
+++ b/internal/homekit/homekit.go
@@ -27,6 +27,7 @@ func Init() {
 			DeviceID      string   `yaml:"device_id"`
 			DevicePrivate string   `yaml:"device_private"`
 			SetupID       string   `yaml:"setup_id"`
+			Image         string   `yaml:"image_stream"`
 			Pairings      []string `yaml:"pairings"`
 		} `yaml:"homekit"`
 	}
@@ -69,6 +70,7 @@ func Init() {
 
 		srv := &server{
 			stream:   id,
+			image:    conf.Image,
 			srtp:     srtp.Server,
 			pairings: conf.Pairings,
 		}
diff --git a/internal/homekit/server.go b/internal/homekit/server.go
index 7aa79f2..75654c6 100644
--- a/internal/homekit/server.go
+++ b/internal/homekit/server.go
@@ -25,6 +25,7 @@ import (
 
 type server struct {
 	stream    string      // stream name from YAML
+	image     string
 	hap       *hap.Server // server for HAP connection and encryption
 	mdns      *mdns.ServiceEntry
 	srtp      *srtp.Server
@@ -142,7 +143,11 @@ func (s *server) SetCharacteristic(conn net.Conn, aid uint8, iid uint64, value a
 func (s *server) GetImage(conn net.Conn, width, height int) []byte {
 	log.Trace().Msgf("[homekit] %s: get image width=%d height=%d", conn.RemoteAddr(), width, height)
 
-	stream := streams.Get(s.stream)
+	streamId := s.stream
+	if(s.image != "") {
+		streamId = s.image
+	}
+	stream := streams.Get(streamId)
 	cons := magic.NewKeyframe()
 
 	if err := stream.AddConsumer(cons); err != nil {
