
diff --git a/examples/homekit/main.go b/examples/homekit/main.go
new file mode 100644
index 0000000..63a5b69
--- /dev/null
+++ b/examples/homekit/main.go
@@ -0,0 +1,27 @@
+package main
+
+import (
+	"github.com/AlexxIT/go2rtc/internal/app"
+	"github.com/AlexxIT/go2rtc/internal/api"
+	"github.com/AlexxIT/go2rtc/internal/streams"
+	"github.com/AlexxIT/go2rtc/internal/rtsp"
+	"github.com/AlexxIT/go2rtc/internal/http"
+	"github.com/AlexxIT/go2rtc/internal/homekit"
+	"github.com/AlexxIT/go2rtc/internal/srtp"
+//	"github.com/AlexxIT/go2rtc/internal/api/ws"
+//	"github.com/AlexxIT/go2rtc/internal/webrtc"
+	"github.com/AlexxIT/go2rtc/pkg/shell" // loop
+)
+
+func main() {
+	app.Init()       // application
+	api.Init()       // /api, /api/homekit
+	streams.Init()   // streams module
+	rtsp.Init()      // for rtsp source stream
+	http.Init()      // for jpeg source stream
+	homekit.Init()   // homekit source
+	srtp.Init()      // SRTP server
+//	ws.Init()        // init WS API endpoint (WebRTC server)
+//	webrtc.Init()    // webrtc source, WebRTC server
+	shell.RunUntilSignal()
+}
